File: validation/scalability/search/run-recommended.py
"""Measure public recommended runs serially with an external ten-minute cap."""
import json
import hashlib
import os
from pathlib import Path
import shutil
import subprocess
import sys
import time

variant = sys.argv[1]
assert variant in {"baseline", "candidate", "candidate_binary_guard", "candidate_fixed_gam"}
cases = sys.argv[2:] or ["friedman_noise", "bank_marketing"]
assert all(case in {"friedman_noise", "bank_marketing"} for case in cases)
output = Path(os.environ.get("AXR_SEARCH_DIR", "~/.cache/autoxplain-scale-0.7.0/search")).expanduser()
source = Path(__file__).resolve().with_name("run-recommended.R")
script_sha256 = hashlib.sha256(source.read_bytes()).hexdigest()
frozen = output / f"run-recommended-{variant}-{script_sha256[:12]}.R"
if frozen.exists():
    assert hashlib.sha256(frozen.read_bytes()).hexdigest() == script_sha256
else:
    shutil.copyfile(source, frozen)
environment = dict(os.environ, OMP_NUM_THREADS="1", OPENBLAS_NUM_THREADS="1", MKL_NUM_THREADS="1")
records = []
library_name = ("baseline-library" if variant == "baseline" else
                "candidate-binary-guard-library" if variant == "candidate_binary_guard" else
                "candidate-library")
library = output / library_name / "AutoXplainR"
installed = {str(path.relative_to(library)): hashlib.sha256(path.read_bytes()).hexdigest()
             for path in sorted(library.rglob("*")) if path.is_file()}
assert installed, "Installed package must be available"
if variant != "baseline":
    manifest = ("candidate-binary-guard-installed-sha256.json" if variant == "candidate_binary_guard"
                else "candidate-final-installed-sha256.json")
    expected = json.loads(source.with_name(manifest).read_text())
    assert installed == expected, "Candidate installation changed after its recorded source snapshot"
(output / f"recommended-{variant}-installed-sha256.json").write_text(
    json.dumps(installed, indent=2)+"\n")
for case in cases:
    destination = output / "recommended" / variant / case
    destination.mkdir(parents=True, exist_ok=True)
    assert not (destination / "summary.json").exists(), "Preserve previous measurements"
    assert not (destination / "fit-timer.json").exists(), "Preserve previous fitting timers"
    command = ["Rscript", str(frozen), case, variant]
    started = time.monotonic()
    with (destination / "run.log").open("w") as stream:
        try:
            result = subprocess.run(command, env=environment, stdout=stream,
                                    stderr=subprocess.STDOUT, timeout=600)
            status, exit_code = "finished", result.returncode
        except subprocess.TimeoutExpired:
            status, exit_code = "timeout", None
    record = dict(case=case, variant=variant, status=status, exit_code=exit_code,
                  process_seconds=time.monotonic()-started, timeout_seconds=600,
                  command=command, script_sha256=script_sha256,
                  native_threads={key: environment[key] for key in
                                  ("OMP_NUM_THREADS", "OPENBLAS_NUM_THREADS", "MKL_NUM_THREADS")})
    records.append(record)
    (output / f"recommended-{variant}-operations.json").write_text(json.dumps(records, indent=2)+"\n")
    print(case, variant, status, exit_code, round(record["process_seconds"], 2), flush=True)
if any(row["status"] != "finished" or row["exit_code"] != 0 for row in records):
    raise SystemExit(1)

File: validation/scalability/search/run-recommended.R
# One public recommended-portfolio fit. The final test set only scores the
# training-selected model and retained alternatives; it never chooses settings.
arguments <- commandArgs(TRUE)
stopifnot(length(arguments) == 2L)
scenario <- match.arg(arguments[[1L]], c("friedman_noise", "bank_marketing"))
variant <- match.arg(arguments[[2L]], c("baseline", "candidate", "candidate_binary_guard", "candidate_fixed_gam"))
output <- Sys.getenv("AXR_SEARCH_DIR", path.expand("~/.cache/autoxplain-scale-0.7.0/search"))
library_path <- file.path(output, if (variant == "baseline") "baseline-library" else if
  (variant == "candidate_binary_guard") "candidate-binary-guard-library" else "candidate-library")
.libPaths(c(library_path, .libPaths()))
library(AutoXplainR)
stopifnot(as.character(packageVersion("AutoXplainR")) == if (variant == "baseline") "0.6.2" else "0.7.0")
cache <- path.expand("~/.cache/autoxplain-stress-0.6.2/benchmark")
case_path <- file.path(cache, "cases", paste0(scenario, ".rds"))
case <- readRDS(case_path)
plan <- jsonlite::read_json(file.path(cache, "benchmark-plan.json"))
declared <- Filter(function(item) identical(item$name, scenario), plan$cases)
stopifnot(length(declared) == 1L, identical(declared[[1L]]$sha256,
  digest::digest(file = case_path, algo = "sha256")))
destination <- file.path(output, "recommended", variant, scenario)
dir.create(destination, recursive = TRUE, showWarnings = FALSE)
stopifnot(!file.exists(file.path(destination, "summary.json")))
writeLines(capture.output(sessionInfo()), file.path(destination, "session-info.txt"))
grids <- if (variant == "candidate_fixed_gam") list(additive = data.frame(
  k = c(5L, 8L, 5L, 8L, 10L), gamma = c(1, 1, 1.4, 1, 1.4),
  select = c(TRUE, TRUE, TRUE, FALSE, TRUE), solver = "gam")) else NULL
control <- tuning_control(grids = grids, fold_ids = case$folds)
record <- list(case = scenario, variant = variant,
  package_version = as.character(packageVersion("AutoXplainR")), library = find.package("AutoXplainR"),
  case_sha256 = declared[[1L]]$sha256, seed = case$fit_seed)
warnings <- character()
started <- proc.time()[["elapsed"]]
tryCatch(withCallingHandlers({
  result <- autoxplain(case$training, "y", test_data = case$evaluation, task = case$task,
    portfolio = "recommended", tuning_control = control, evaluation_role = "test",
    explain = FALSE, seed = case$fit_seed)
  record$fit_seconds <- proc.time()[["elapsed"]] - started
  # Persist the call timer before prediction, summarization, or serialization of
  # complex evidence objects. A later bookkeeping failure must not lose it.
  jsonlite::write_json(list(case = scenario, variant = variant,
    fit_seconds = record$fit_seconds, package_version = record$package_version),
    file.path(destination, "fit-timer.json"), auto_unbox = TRUE, pretty = TRUE, digits = 16)
  saveRDS(list(result = result, case = case, record = record), file.path(destination, "result.rds"))
  predictions <- lapply(names(result$models), function(id) predict(result, case$evaluation, model = id))
  names(predictions) <- names(result$models)
  scores <- lapply(predictions, function(p) {
    y <- case$evaluation$y
    stopifnot(length(p) == length(y), all(is.finite(p)))
    if (case$task == "regression") return(list(rmse = sqrt(mean((y - p)^2)), mae = mean(abs(y - p))))
    event <- as.integer(y == "yes")
    stopifnot(all(p >= 0 & p <= 1))
    bounded <- pmax(1e-15, pmin(1 - 1e-15, p))
    list(log_loss = -mean(event * log(bounded) + (1 - event) * log1p(-bounded)),
      brier = mean((p - event)^2),
      auc = (sum(rank(p)[event == 1L]) - sum(event) * (sum(event) + 1) / 2) /
        (sum(event) * sum(1 - event)))
  })
  record$metrics <- scores
  record$primary <- result$provenance$primary_model_id
  record$selected_configuration <- result$tuning$final_configuration
  record$model_count <- length(result$models)
  record$configuration_count <- nrow(result$tuning$candidates)
  record$configuration_status <- as.list(table(result$tuning$candidates$status))
  record$candidates <- result$tuning$candidates[, c("configuration_id", "family",
    "hyperparameters", "cv_score", "folds_completed", "selected", "status")]
  record$input_policy <- result$tuning$input_policy
  record$refit <- result$tuning$refit
  folds <- result$tuning$fold_scores
  record$failed_folds <- folds[!is.finite(folds$score), c("configuration_id", "fold",
    "optimization_status", "optimization_message", "warning", "error")]
  families <- result$tuning$candidates$family[match(folds$configuration_id,
    result$tuning$candidates$configuration_id)]
  record$fold_fit_seconds_by_family <- aggregate(folds$elapsed_ms / 1000,
    list(family = families), sum)
  record$status <- "ok"
  saveRDS(list(result = result, case = case, record = record), file.path(destination, "result.rds"))
  saveRDS(predictions, file.path(destination, "predictions.rds"))
}, warning = function(w) {
  warnings <<- c(warnings, conditionMessage(w))
  invokeRestart("muffleWarning")
}), error = function(e) {
  record$status <<- "failed"
  record$error <<- conditionMessage(e)
})
record$elapsed_seconds <- proc.time()[["elapsed"]] - started
record$warnings <- as.list(unname(warnings))
jsonlite::write_json(record, file.path(destination, "summary.json"),
  pretty = TRUE, auto_unbox = TRUE, digits = 16, null = "null", na = "null")
cat(scenario, variant, record$status, record$elapsed_seconds, "seconds\n")
if (record$status != "ok") quit(status = 1L)
